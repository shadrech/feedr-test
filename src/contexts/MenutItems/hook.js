import React from 'react';
import defaultData from './items';

export default function useMenuItems() {
  const [unselectedItems, setUnselectedItems] = React.useState(defaultData);
  const [selectedItems, setSelectedItems] = React.useState([]);

  const selectItem = (sourceIndex, destIndex) => {
    const unselected = [...unselectedItems];
    const item = unselected.splice(sourceIndex, 1);
    const selected = [...selectedItems.slice(0, destIndex), item[0], ...selectedItems.slice(destIndex)];
    setUnselectedItems(unselected);
    setSelectedItems(selected);
  }

  const unselectItem = (sourceIndex, destIndex) => {
    const selected = [...selectedItems];
    const item = selected.splice(sourceIndex, 1);
    const unselected = [...unselectedItems.slice(0, destIndex), item[0], ...unselectedItems.slice(destIndex)];
    setUnselectedItems(unselected);
    setSelectedItems(selected);
  }

  const removeSelectedItem = index => {
    const selected = [...selectedItems];
    const item = selected.splice(index, 1);
    const unselected = [...unselectedItems, item[0]];
    setUnselectedItems(unselected);
    setSelectedItems(selected);
  }

  return { unselectedItems, selectedItems, selectItem, unselectItem, removeSelectedItem };
}
