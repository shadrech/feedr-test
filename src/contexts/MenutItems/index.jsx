import React from 'react';
import useMenuItems from './hook';

const Context = React.createContext()

export const useContext = () => {
  const context = React.useContext(Context)

  if (!context) throw new Error('Cannot use useMenuItemContext when not within a MenuItemContextProvider');

  return context;
}

export function ContextProvider(props) {
  const value = useMenuItems()

  return <Context.Provider {...props} value={value} />
}
