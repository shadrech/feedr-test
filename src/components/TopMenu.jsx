import React from 'react'
import * as menuItems from '../contexts/MenutItems'

export const TopMenu = () => {
  const { selectedItems } = menuItems.useContext()
  const dietaries = selectedItems.reduce((prev, curr) => {
    curr.dietaries.forEach(diet => {
      prev[diet] = !!prev[diet] ? ++prev[diet] : 1
    })
    return prev;
  }, {})

  return (
    <div className="menu-summary">
      <div className="container">
        <div className="row">
          <div className="col-6 menu-summary-left">
            <span>{selectedItems.length} items</span>
          </div>
          <div className="col-6 menu-summary-right">
            {dietaries && Object.keys(dietaries).map(diet => (
              <span key={diet}>
                {dietaries[diet]}x <span className="dietary">{diet}</span>
              </span>
            ))}
          </div>
        </div>
      </div>
    </div>
  )
}
