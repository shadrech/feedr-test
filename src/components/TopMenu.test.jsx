import React from 'react'
import { shallow, configure } from 'enzyme';
import * as mi from '../contexts/MenutItems';
import { TopMenu } from './TopMenu';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('<TopMenu />', () => {

  it('renders component with correct stats', () => {
    jest.spyOn(mi, 'useContext').mockReturnValue({
      selectedItems: [{
        dietaries: ['gf', 'ab']
      }, {
        dietaries: ['gf', 'dc']
      }]
    })
    const component = shallow(<TopMenu />)
    expect(component.html()).toMatchSnapshot()
  })

})
