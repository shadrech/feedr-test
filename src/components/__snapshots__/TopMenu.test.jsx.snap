// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopMenu /> renders component with correct stats 1`] = `"<div class=\\"menu-summary\\"><div class=\\"container\\"><div class=\\"row\\"><div class=\\"col-6 menu-summary-left\\"><span>2 items</span></div><div class=\\"col-6 menu-summary-right\\"><span>2x <span class=\\"dietary\\">gf</span></span><span>1x <span class=\\"dietary\\">ab</span></span><span>1x <span class=\\"dietary\\">dc</span></span></div></div></div></div>"`;
